README.md
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://raw.githubusercontent.com/fewieden/MMM-voice/master/LICENSE) [![Build Status](https://travis-ci.org/fewieden/MMM-voice.svg?branch=master)](https://travis-ci.org/fewieden/MMM-voice) [![Code Climate](https://codeclimate.com/github/fewieden/MMM-voice/badges/gpa.svg?style=flat)](https://codeclimate.com/github/fewieden/MMM-voice) [![Known Vulnerabilities](https://snyk.io/test/github/fewieden/mmm-voice/badge.svg)](https://snyk.io/test/github/fewieden/mmm-voice) [![Documentation](https://img.shields.io/badge/Documentation-master-brightgreen.svg)](https://fewieden.github.io/MMM-voice/)

# MMM-voice

Offline Voice Recognition Module for MagicMirror<sup>2</sup>.

## Information

This voice recognition works offline. To protect your privacy, no one will record what's going on in your room all day long.
So keep in mind that there is no huge server farm, that handles your voice commands. The raspberry is just a small device and this is a cpu intensive task.
Also the dictionary has only the words specified by the modules, so there is a chance for false positives.

If you can live with latency, bugged detections and want to have data privacy, feel free to use this module.

## Dependencies

* An installation of [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror)
* Packages: bison libasound2-dev autoconf automake libtool python-dev swig python-pip
* [SphinxBase](https://github.com/cmusphinx/sphinxbase)
* [PocketSphinx](https://github.com/cmusphinx/pocketsphinx)
* A microphone
* npm
* [PocketSphinx-continuous](https://www.npmjs.com/package/pocketsphinx-continuous)
* [lmtool](https://www.npmjs.com/package/lmtool)
* [fs-extra](https://www.npmjs.com/package/fs-extra)
* [lodash](https://www.npmjs.com/package/lodash)

## Installation

1. Clone this repo into `~/MagicMirror/modules` directory.
1. Run command `bash dependencies.sh` in `~/MagicMirror/modules/MMM-voice/installers` directory, to install all dependencies. This will need a couple of minutes.
1. Configure your `~/MagicMirror/config/config.js`:

    ```
    {
        module: 'MMM-voice',
        position: 'bottom_bar',
        config: {
            microphone: 1,
            standByMethod: 'TVSERVICE',
            ...		
        }
    }
    ```

## Config Options

| **Option** | **Default** | **Description** |
| --- | --- | --- |
| `microphone` | REQUIRED | Id of microphone shown in the installer. |
| `keyword` | `'MAGIC MIRROR'` | Keyword the mirror starts to listen. IMPORTANT: Only UPPERCASE Letters |
| `timeout` | `15` | time the keyword should be active without saying something |
| `standByMethod` | `'TVSERVICE'` | `'TVSERVICE'` uses the tvservice command available on Raspberry pi to turn off the HDMI monitor source (default). <br>`'DPMS'` uses the exec DMPS command to turn off the monitor source (not on pi, or not hdmi). <br>`'VCGENCMD'` uses the vcgencmd to turn on/off video output. <br>`'HIDE'` hides all module contents, if display is on EnergyStar device that shows ugly 'no signal' screen for the other choices. |

## Usage

You need to say your KEYWORD (Default: MAGIC MIRROR), when the KEYWORD is recognized the microphone will start to flash and as long as the microphone is flashing (timeout config option) the mirror will recognize COMMANDS or MODES (Keep in mind that the recognition will take a while, so when you say your COMMAND right before the microphone stops flashing the COMMAND will probably not recognized).

Mode of this module: `VOICE`

COMMANDS:

* HIDE MODULES
* SHOW MODULES
* WAKE UP
* GO TO SLEEP
* OPEN HELP
* CLOSE HELP

### Select Mode

To select a MODE, the specific MODE has to be the first word of a COMMAND or right after the KEYWORD, when the microphone stopped flashing.

## Supported modules

List of all supported modules in the [Wiki](https://github.com/fewieden/MMM-voice/wiki/Supported-Modules).

## Developer

* `npm run lint` - Lints JS and CSS files.
* `npm run docs` - Generates documentation.

### Documentation

The documentation can be found [here](https://doclets.io/fewieden/MMM-voice/master)

### Developers Guide

If you want to support your own module, check out the [Guide](DEVELOPER.md) and add it to the [Wiki](https://github.com/fewieden/MMM-voice/wiki/Supported-Modules).


CHANGELOG.md
# MMM-voice Changelog

## [2.0.0]

### Added

* Nunjuck templates
* Multiple stand by methods

### Changed

* Updated config files
* async/await on the node side of the module.
* fs-extra for promised based filesystem access.
* Documentation is now hosted on github pages instead of doclets.io (unmaintained).
* eslint recommended instead of airbnb ruleset.

### Removed

* Bytes.js (easteregg)
* Modals (help)

## [1.0.2]

### Fixed

* Restored context for handleData, logDebug and logError
* DeprecationWarning: Calling async function without callback (noop function)

## [1.0.1]

### Added

* Code linter
* Documentation
* [Doclets.io](https://doclets.io/fewieden/MMM-voice/master) integration
* Contributing guidelines
* Issue template
* Pull request template
* Gitignore

### Changed

* Dependency manipulation exchanged to fork

## [1.0.0]

Initial version


node_helper.js
/**
 * @file node_helper.js
 *
 * @author fewieden
 * @license MIT
 *
 * @see  https://github.com/fewieden/MMM-voice
 */

/**
 * @external pocketsphinx-continuous
 * @see https://github.com/fewieden/pocketsphinx-continuous-node
 */
const Psc = require('pocketsphinx-continuous');

/**
 * @external fs-extra
 * @see https://www.npmjs.com/package/fs-extra
 */
const fs = require('fs-extra');

/**
 * @external lodash
 * @see https://www.npmjs.com/package/lodash
 */
const _ = require('lodash');

/**
 * @external child_process
 * @see https://nodejs.org/api/child_process.html
 */
const exec = require('child_process').exec;

/**
 * @external lmtool
 * @see https://www.npmjs.com/package/lmtool
 */
const lmtool = require('lmtool');

/**
 * @external node_helper
 * @see https://github.com/MichMich/MagicMirror/blob/master/modules/node_modules/node_helper/index.js
 */
const NodeHelper = require('node_helper');

/**
 * @module node_helper
 * @description Backend for the module to query data from the API providers.
 *
 * @requires external:pocketsphinx-continuous
 * @requires external:fs
 * @requires external:child_process
 * @requires external:lmtool
 * @requires external:node_helper
 */
module.exports = NodeHelper.create({

    /** @member {boolean} listening - Flag to indicate listen state. */
    listening: false,

    /** @member {boolean|string} mode - Contains active module mode. */
    mode: false,

    /** @member {boolean} help - Flag to toggle help modal. */
    help: false,

    /** @member {string[]} words - List of all words that are registered by the modules. */
    words: [],

    /** @member {Object} standBy - Mapping of stand by methods with commands. */
    standBy: {
        TVSERVICE: {
            hardware: true,
            show: '/opt/vc/bin/tvservice -p && sudo chvt 6 && sudo chvt 7',
            hide: '/opt/vc/bin/tvservice -o'
        },
        DPMS: {
            hardware: true,
            show: 'xset dpms force on',
            hide: 'xset dpms force off'
        },
        VCGENCMD: {
            hardware: true,
            show: 'vcgencmd display_power 1',
            hide: 'vcgencmd display_power 0'
        },
        HIDE: {
            hardware: false
        }
    },

    /**
     * @function socketNotificationReceived
     * @description Receives socket notifications from the module.
     * @async
     * @override
     *
     * @param {string} notification - Notification name
     * @param {*} payload - Detailed payload of the notification.
     */
    async socketNotificationReceived(notification, payload) {
        if (notification === 'START') {
            /** @member {Object} config - Module config. */
            this.config = payload.config;
            /** @member {number} time - Time to listen after keyword. */
            this.time = this.config.timeout * 1000;
            /** @member {Object} modules - List of modules with their modes and commands. */
            this.modules = payload.modules;

            this.fillWords();
            await this.checkFiles();
        }
    },

    /**
     * @function fillwords
     * @description Sets {@link node_helper.words} with all needed words for the registered
     * commands by the modules. This list has unique items and is sorted by alphabet.
     */
    fillWords() {
        let allWords = _.split(this.config.keyword, ' ');

        for (const module of this.modules) {
            const mode = _.split(module.mode, ' ');
            allWords = _.concat(allWords, mode);
            for (const sentence of module.sentences) {
                const words = _.split(sentence, ' ');
                allWords = _.concat(allWords, words);
            }
        }

        this.words = _.sortBy(_.uniq(allWords));
    },

    /**
     * @function checkFiles
     * @description Checks if words.json exists or has different entries as this.word.
     * @async
     */
    async checkFiles() {
        console.log(`${this.name}: Checking files.`);
        const file = 'modules/MMM-voice/words.json';
        const exists = await fs.pathExists(file);

        if (exists) {
            const words = await fs.readJson(file);
            if (_.isEqual(this.words, words)) {
                return this.startPocketsphinx();
            }
        }

        this.generateDicLM();
    },

    /**
     * @function generateDicLM
     * @description Generates new Dictionairy and Language Model.
     * @async
     */
    async generateDicLM() {
        console.log(`${this.name}: Generating dictionairy and language model.`);

        await fs.writeJson('modules/MMM-voice/words.json', this.words);

        lmtool(this.words, async (err, filename) => {
            if (err) {
                this.sendSocketNotification('ERROR', 'Couldn\'t create necessary files!');
            } else {
                await fs.move(`${filename}.dic`, 'modules/MMM-voice/MMM-voice.dic', {overwrite: true});
                await fs.move(`${filename}.lm`, 'modules/MMM-voice/MMM-voice.lm', {overwrite: true});

                this.startPocketsphinx();

                await fs.remove(`${filename}.log_pronounce`);
                await fs.remove(`${filename}.sent`);
                await fs.remove(`${filename}.vocab`);
                await fs.remove(`TAR${filename}.tgz`);
            }
        });
    },

    /**
     * @function startPocketsphinx
     * @description Starts Pocketsphinx binary.
     */
    startPocketsphinx() {
        console.log(`${this.name}: Starting pocketsphinx.`);

        this.ps = new Psc({
            setId: this.name,
            verbose: true,
            microphone: this.config.microphone
        });

        this.ps.on('data', this.handleData.bind(this));

        if (this.config.debug) {
            this.ps.on('debug', this.logDebug.bind(this));
        }

        this.ps.on('error', this.logError.bind(this));

        this.sendSocketNotification('READY');
    },

    /**
     * @function handleData
     * @description Helper method to handle recognized data.
     *
     * @param {string} data - Recognized data
     */
    handleData(data) {
        if (typeof data === 'string') {
            if (this.config.debug) {
                console.log(`${this.name} has recognized: ${data}`);
                this.sendSocketNotification('DEBUG', data);
            }

            if (_.includes(data, this.config.keyword) || this.listening) {
                this.listening = true;
                this.sendSocketNotification('LISTENING');

                if (this.timer) {
                    clearTimeout(this.timer);
                }

                this.timer = setTimeout(() => {
                    this.listening = false;
                    this.sendSocketNotification('SLEEPING');
                }, this.time);
            } else {
                return;
            }

            let cleanData = this.cleanData(data);

            for (const module of this.modules) {
                const i = _.indexOf(cleanData, module.mode);
                if (i === 0) {
                    this.mode = module.mode;
                    cleanData = _.trim(cleanData.substr(i + module.mode.length));
                    break;
                }
            }

            if (this.mode) {
                this.sendSocketNotification('VOICE', { mode: this.mode, sentence: cleanData });
                if (this.mode === 'VOICE') {
                    this.checkCommands(cleanData);
                }
            }
        }
    },

    /**
     * @function logDebug
     * @description Logs debug information into debug log file.
     * @async
     *
     * @param {string} data - Debug information
     */
    async logDebug(data) {
        try {
            await fs.appendFile('modules/MMM-voice/debug.log', `${data}\n`);
        } catch(e) {
            console.log(`${this.name}: Couldn't save debug information to log file!`, e);
        }
    },

    /**
     * @function logError
     * @description Logs error information into error log file.
     * @async
     *
     * @param {string} error - Error information
     */
    async logError(error) {
        if (error) {
            try {
                await fs.appendFile('modules/MMM-voice/error.log', `${error}\n`);
            } catch(e) {
                console.log(`${this.name}: Couldn't save error to log file!`, e);
            }

            this.sendSocketNotification('ERROR', error);
        }
    },

    /**
     * @function cleanData
     * @description Removes prefix/keyword and multiple spaces.
     *
     * @param {string} data - Recognized data to clean.
     * @returns {string} Cleaned data
     */
    cleanData(data) {
        const i = _.indexOf(data, this.config.keyword);

        if (i !== -1) {
            data = data.substr(i + this.config.keyword.length);
        }

        return _.trim(_.replace(data, / {2,}/g, ' '));
    },

    /**
     * @function handleStandBy
     * @description Handles stand by with various methods.
     *
     * @param {string} method - Stand by method to use.
     * @param {string} action - Action to perform.
     */
    handleStandBy(method, action) {
        if (this.standBy.hasOwnProperty(method)) {
            if (this.standBy[method].hardware) {
                exec(this.standBy[method][action], null);
            }

            this.sendSocketNotification('STAND_BY_ACTION', {
                type: action,
                hardware: this.standBy[method].hardware
            });
        }
    },

    /**
     * @function checkCommands
     * @description Checks for commands of voice module
     * @param {string} data - Recognized data
     */
    checkCommands(data) {
        if (/(WAKE)/g.test(data) && /(UP)/g.test(data)) {
            this.handleStandBy(this.config.standByMethod.toUpperCase(), 'show');
        } else if (/(GO)/g.test(data) && /(SLEEP)/g.test(data)) {
            this.handleStandBy(this.config.standByMethod.toUpperCase(), 'hide');
        } else if (/(SHOW)/g.test(data) && /(MODULES)/g.test(data)) {
            this.sendSocketNotification('SHOW');
        } else if (/(HIDE)/g.test(data) && /(MODULES)/g.test(data)) {
            this.sendSocketNotification('HIDE');
        } else if (/(HELP)/g.test(data)) {
            if (/(CLOSE)/g.test(data) || (this.help && !/(OPEN)/g.test(data))) {
                this.sendSocketNotification('CLOSE_HELP');
                this.help = false;
            } else if (/(OPEN)/g.test(data) || (!this.help && !/(CLOSE)/g.test(data))) {
                this.sendSocketNotification('OPEN_HELP');
                this.help = true;
            }
        }
    }
});


installers/dependencies.sh
#!/usr/bin/env bash

# Magic Mirror
# Module: MMM-voice
#
# By fewieden https://github.com/fewieden/MMM-voice
# MIT Licensed.

echo -e "\e[0m"
echo '       `/ymNMNds-'
echo '      :mMMMMMMMMMy`'
echo '     .NMMMMMMMMMMMh'
echo '     /MMMMMMMMMMMMN'
echo '     /MMMMMMMMMMMMN                       ``    `.:+oo/.         ``       ``````       `````.`'
echo '     /MMMMMMMMMMMMN           .-:+oy` +syhd-  `/hNMMMMMNs`   `osyhy     -ohmNNmh/` +syhhdmmNM:'
echo '     /MMMMMMMMMMMMN           dMMMMm /MMMMs  :dMMMMMMMMMMs   +MMMM/   -yNMMMMMMMM+.MMMMMMMMMM`'
echo '     /MMMMMMMMMMMMN           mMMMMy dMMMm``sMMMMMMNNMMMMN   dMMMm   +NMMMMMNNMd/ oMMMMmdhyys'
echo '     /MMMMMMMMMMMMN           NMMMM/:MMMM-`yMMMMNy:--dMMMM` .MMMM+  sMMMMMh:--:`  dMMMM. ```'
echo ' `-- /MMMMMMMMMMMMN `-.       MMMMM`dMMMs sMMMMN/    /MMMm  oMMMN` +MMMMN+       .MMMMNshdd+'
echo ' /MM`/MMMMMMMMMMMMN +Nm       MMMMd:MMMm`:MMMMN-     /MMM+  mMMMs `NMMMM:        oMMMMMMMMN/'
echo ' /MM`/MMMMMMMMMMMMN oMm      `MMMModMMM/ yMMMM/      dMMh` -MMMM- /MMMMs         mMMMmo/:-.`'
echo ' :MM.:MMMMMMMMMMMMm oMm      .MMMMoMMMh  mMMMM.    `sMMh`  sMMMd  +MMMM+    `.  -MMMMh/+osy'
echo ' `NMs yMMMMMMMMMMN:`mMs      :MMMNdMMM-  dMMMMh:--+dMMs`   NMMMo  -MMMMm/--+my  oMMMMMMMMMN'
echo '  :NMs./dNMMMMMNy-:dMh`      oMMMNMMMy   /MMMMMMNNMMd/    :MMMM-   oMMMMMNNMMh  mMMMMMMMMMd'
echo '   .yNNy///+++//+dNm+`       hNmdhyso.    /mMMMMMNh/`     yNdhs     :hmNNmho-` -Nmdhso+/:-.'
echo '     .ohmNNmmmNNmy/`         .```          `-/+/-`        .``         `````    `.``'
echo '        `.-NMy-``'
echo '           NMo'
echo '     ......NMy.....`'
echo '     mMMMMMMMMMMMMMo'
echo -e "\e[0m"


# installing packages
echo -e "\e[96m[STEP 1/5] Installing Packages\e[90m"
if sudo apt-get install bison libasound2-dev autoconf automake libtool python-dev swig python-pip -y ;
then
    echo -e "\e[32m[STEP 1/5] Installing Packages | Done\e[0m"
else
	echo -e "\e[31m[STEP 1/5] Installing Packages | Failed\e[0m"
	exit;
fi


# installing sphinxbase
echo -e "\e[96m[STEP 2/5] Installing sphinxbase\e[90m"
cd ~
if [ ! -d "$HOME/sphinxbase" ] ;
then
    if ! git clone https://github.com/cmusphinx/sphinxbase.git ;
    then
        echo -e "\e[31m[STEP 2/5] Installing sphinxbase | Failed\e[0m"
        exit;
    fi
fi

cd sphinxbase
if ! git pull ;
then
    echo -e "\e[31m[STEP 2/5] Installing sphinxbase | Failed\e[0m"
    exit;
fi

./autogen.sh
./configure --enable-fixed
make
sudo make install
echo -e "\e[32m[STEP 2/5] Installing sphinxbase | Done\e[0m"


# installing pocketsphinx
echo -e "\e[96m[STEP 3/5] Installing pocketsphinx\e[90m"
cd ~
if [ ! -d "$HOME/pocketsphinx" ] ;
then
    if ! git clone https://github.com/cmusphinx/pocketsphinx.git ;
    then
        echo -e "\e[31m[STEP 3/5] Installing pocketsphinx | Failed\e[0m"
        exit;
    fi
fi

cd pocketsphinx
if ! git pull ;
then
    echo -e "\e[31m[STEP 3/5] Installing pocketsphinx | Failed\e[0m"
    exit;
fi

./autogen.sh
./configure
make
sudo make install
echo -e "\e[32m[STEP 3/5] Installing pocketsphinx | Done\e[0m"


# exporting paths
echo -e "\e[96m[STEP 4/5] Exporting paths\e[0m"
echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib" >> ~/.bashrc
echo "export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:/usr/local/lib/pkgconfig" >> ~/.bashrc
echo -e "\e[32m[STEP 4/5] Exporting paths |  Done\e[0m"


# installing npm dependencies
echo -e "\e[96m[STEP 5/5] Installing npm dependencies\e[90m"
cd ~/MagicMirror/modules/MMM-voice
if npm install --productive;
then
    echo -e "\e[32m[STEP 5/5] Installing npm dependencies | Done\e[0m"
else
    echo -e "\e[31m[STEP 5/5] Installing npm dependencies | Failed\e[0m"
    exit;
fi


# displaying audio devices
echo -e "\e[96m[INFO] Possible Audio Devices to set in config.js\n\e[0m"
cat /proc/asound/cards


MMM-voice.js
/**
 * @file MMM-voice.js
 *
 * @author fewieden
 * @license MIT
 *
 * @see  https://github.com/fewieden/MMM-voice
 */

/* global Module Log MM */

/**
 * @external Module
 * @see https://github.com/MichMich/MagicMirror/blob/master/js/module.js
 */

/**
 * @external Log
 * @see https://github.com/MichMich/MagicMirror/blob/master/js/logger.js
 */

/**
 * @external MM
 * @see https://github.com/MichMich/MagicMirror/blob/master/js/main.js
 */

/**
 * @module MMM-voice
 * @description Frontend for the module to display data.
 *
 * @requires external:Module
 * @requires external:Log
 * @requires external:MM
 */
Module.register('MMM-voice', {

    /** @member {string} icon - Microphone icon. */
    icon: 'fa-microphone-slash',
    /** @member {boolean} pulsing - Flag to indicate listening state. */
    pulsing: true,
    /** @member {boolean} help - Flag to switch between render help or not. */
    help: false,

    /**
     * @member {Object} voice - Defines the default mode and commands of this module.
     * @property {string} mode - Voice mode of this module.
     * @property {string[]} sentences - List of voice commands of this module.
     */
    voice: {
        mode: 'VOICE',
        sentences: [
            'HIDE MODULES',
            'SHOW MODULES',
            'WAKE UP',
            'GO TO SLEEP',
            'OPEN HELP',
            'CLOSE HELP',
            'NEXT PAGE'
        ]
    },

    /** @member {Object[]} modules - Set of all modules with mode and commands. */
    modules: [],

    /**
     * @member {Object} defaults - Defines the default config values.
     * @property {int} timeout - Seconds to active listen for commands.
     * @property {string} keyword - Keyword to activate active listening.
     * @property {boolean} debug - Flag to enable debug information.
     */
    defaults: {
        timeout: 15,
        keyword: 'MAGIC MIRROR',
        debug: false
    },

    /**
     * @function start
     * @description Sets mode to initialising.
     * @override
     */
    start() {
        Log.info(`Starting module: ${this.name}`);
        this.mode = this.translate('INIT');
        this.modules.push(this.voice);
        Log.info(`${this.name} is waiting for voice command registrations.`);
    },

    /**
     * @function getStyles
     * @description Style dependencies for this module.
     * @override
     *
     * @returns {string[]} List of the style dependency filepaths.
     */
    getStyles() {
        return ['font-awesome.css', 'MMM-voice.css'];
    },

    /**
     * @function getTranslations
     * @description Translations for this module.
     * @override
     *
     * @returns {Object.<string, string>} Available translations for this module (key: language code, value: filepath).
     */
    getTranslations() {
        return {
            en: 'translations/en.json',
            de: 'translations/de.json',
            id: 'translations/id.json'
        };
    },

    /**
     * @function getDom
     * @description Creates the UI as DOM for displaying in MagicMirror application.
     * @override
     *
     * @returns {Element}
     */
    getDom() {
        const wrapper = document.createElement('div');
        const voice = document.createElement('div');
        voice.classList.add('small', 'align-left');

        const icon = document.createElement('i');
        icon.classList.add('fa', this.icon, 'icon');
        if (this.pulsing) {
            icon.classList.add('pulse');
        }
        voice.appendChild(icon);

        const modeSpan = document.createElement('span');
        modeSpan.innerHTML = this.mode;
        voice.appendChild(modeSpan);
        if (this.config.debug) {
            const debug = document.createElement('div');
            debug.innerHTML = this.debugInformation;
            voice.appendChild(debug);
        }

        const modules = document.querySelectorAll('.module');
        for (let i = 0; i < modules.length; i += 1) {
            if (!modules[i].classList.contains(this.name)) {
                if (this.help) {
                    modules[i].classList.add(`${this.name}-blur`);
                } else {
                    modules[i].classList.remove(`${this.name}-blur`);
                }
            }
        }

        if (this.help) {
            voice.classList.add(`${this.name}-blur`);
            const modal = document.createElement('div');
            modal.classList.add('modal');
            this.appendHelp(modal);
            wrapper.appendChild(modal);
        }

        wrapper.appendChild(voice);

        return wrapper;
    },

    /**
     * @function notificationReceived
     * @description Handles incoming broadcasts from other modules or the MagicMirror core.
     * @override
     *
     * @param {string} notification - Notification name
     * @param {*} payload - Detailed payload of the notification.
     */
    notificationReceived(notification, payload) {
        if (notification === 'DOM_OBJECTS_CREATED') {
            this.sendSocketNotification('START', { config: this.config, modules: this.modules });
        } else if (notification === 'REGISTER_VOICE_MODULE') {
            if (Object.prototype.hasOwnProperty.call(payload, 'mode') && Object.prototype.hasOwnProperty.call(payload, 'sentences')) {
                this.modules.push(payload);
            }
        }
    },

    /**
     * @function socketNotificationReceived
     * @description Handles incoming messages from node_helper.
     * @override
     *
     * @param {string} notification - Notification name
     * @param {*} payload - Detailed payload of the notification.
     */
    socketNotificationReceived(notification, payload) {
        if (notification === 'READY') {
            this.icon = 'fa-microphone';
            this.mode = this.translate('NO_MODE');
            this.pulsing = false;
        } else if (notification === 'LISTENING') {
            this.pulsing = true;
        } else if (notification === 'SLEEPING') {
            this.pulsing = false;
        } else if (notification === 'ERROR') {
            this.mode = notification;
        } else if (notification === 'VOICE') {
            for (let i = 0; i < this.modules.length; i += 1) {
                if (payload.mode === this.modules[i].mode) {
                    if (this.mode !== payload.mode) {
                        this.help = false;
                        this.sendNotification(`${notification}_MODE_CHANGED`, { old: this.mode, new: payload.mode });
                        this.mode = payload.mode;
                    }
                    if (this.mode !== 'VOICE') {
                        this.sendNotification(`${notification}_${payload.mode}`, payload.sentence);
                    }
                    break;
                }
            }
        } else if (notification === 'BYTES') {
            this.sendNotification('MMM-TTS', payload);
        } else if (notification === 'HIDE') {
            MM.getModules().enumerate((module) => {
                module.hide(1000);
            });
        } else if (notification === 'SHOW') {
            MM.getModules().enumerate((module) => {
                module.show(1000);
            });
        } else if (notification === 'OPEN_HELP') {
            this.help = true;
        } else if (notification === 'CLOSE_HELP') {
            this.help = false;
     // added for MMM-pages to show next page // 
        } else if (notification === 'NEXT_PAGE') {
			this.sendNotification('PAGE_INCREMENT');
  //////////////////////////////////////////////////////
        } else if (notification === 'DEBUG') {
            this.debugInformation = payload;
        }
        this.updateDom(300);
    },

    /**
     * @function appendHelp
     * @description Creates the UI for the voice command SHOW HELP.
     *
     * @param {Element} appendTo - DOM Element where the UI gets appended as child.
     */
    appendHelp(appendTo) {
        const title = document.createElement('h1');
        title.classList.add('medium');
        title.innerHTML = `${this.name} - ${this.translate('COMMAND_LIST')}`;
        appendTo.appendChild(title);

        const mode = document.createElement('div');
        mode.innerHTML = `${this.translate('MODE')}: ${this.voice.mode}`;
        appendTo.appendChild(mode);

        const listLabel = document.createElement('div');
        listLabel.innerHTML = `${this.translate('VOICE_COMMANDS')}:`;
        appendTo.appendChild(listLabel);

        const list = document.createElement('ul');
        for (let i = 0; i < this.voice.sentences.length; i += 1) {
            const item = document.createElement('li');
            item.innerHTML = this.voice.sentences[i];
            list.appendChild(item);
        }
        appendTo.appendChild(list);
    }
});
